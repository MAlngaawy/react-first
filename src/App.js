/*
 1 ==> Props are used to pass data from component to component
 2 ==> props are only passed from top to bottom in react's component tree
 3 ==> there is no way in react to set props == Props are read-only.
 4 ==> props are only used to pass data from one component to another component React, but only from parent to child components down the component tree.
 5 ==> lifecycle methods Just "componentDidMount()" method use to ({fech Date})
*/
import React, {Component} from 'react';
import {useState} from 'react';
import './App.css';

const DEFAULT_QUERY = 'javascript';
const DEFAULT_HPP = '10'; // to control the num of hits 20 or 100 or 50

const PATH_BASE = 'https://hn.algolia.com/api/v1';
const PATH_SEARCH = '/search';
const PARAM_SEARCH = 'query=';
const PARAM_PAGE = 'page=';
const PARAM_HPP = 'hitsPerPage='; // to control the num of hits 20 or 100 or 50

// URL
const url =  `${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${DEFAULT_QUERY}`;

const Test = () => {
  const [value, setValue] = useState('Hello World') // Used Hookes Here instead of useState Function
  const handleChange = event => setValue(event.target.value);

  return (
    <div>
      <input type='text' value={value} onChange={handleChange} />
      <p>
        <strong>Output: </strong> {value}
      </p>
    </div>
  )
}


// now we use (functional stateless component) instead of (ES6 class components)
const Search = ({value, onChange,onSubmit, children}) => // ({this.Props})
      <form onSubmit={onSubmit}>
        <input
          type="text"
          value={value}
          onChange={onChange}
        />
        <button type='submit'>
          {children}
        </button>
      </form>

const Loading = () =>
<div> 
<h1 className='loading'>Waite MotherFuckr</h1>
</div>

const Table = ({list, onDismiss, plusPoints}) => 
      <div className='table'>
        {list.map(item => 
          <div key={item.objectID} className='table-row'>
          <span style={{ width: '40%' }}> 
              <a href={item.url} target='_blank'>{item.title}</a>
            </span>
            <span style={{ width: '30%' }}> {item.author} </span>
            <span style={{ width: '10%' }}> {item.num_comments} </span>
            <span style={{ width: '10%' }}> {item.points} </span>
            <span style={{ width: '10%' }}>
            <Button onClick={() => onDismiss(item.objectID)} className='button-inline'>
              onDismiss
            </Button>
            {/* <Button onClick={() => plusPoints(list.indexOf(item))}>plusPoints</Button> */}
            </span>
          </div>
        )}
      </div>

const Button = ({onClick, className='', children}) =>
      <button
        onClick={onClick}
        className={className}
        type='button'
      >
        {children}
      </button>


class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      result: null,
      searchTerm: DEFAULT_QUERY,
    };

    this.fetchSearchTopStories = this.fetchSearchTopStories.bind(this)
    this.setSearchTopStories = this.setSearchTopStories.bind(this)
    this.onSearchChange = this.onSearchChange.bind(this)
    this.onSearchSubmit = this.onSearchSubmit.bind(this)
    this.onDismiss = this.onDismiss.bind(this)
  }

  // it's for change the value of searchTerm depending on the input field value
  setSearchTopStories(result) {
    // let's show all pages instead of the just last page
    const {hits, page} = result
    // if the page is 0 return rmptey array ==== if the page not equal 0 return the current hits in result
    const oldHits = page !== 0 ? this.state.result.hits : [];
    // concat all the old hits from the previose const with the hits witch comes back in result from last fetch
    const updatedHits = [...oldHits, ...hits];
    //Update the state
    this.setState({
      result: {hits: updatedHits, page}
    })
  }

  fetchSearchTopStories(searchTerm, page = 0) {
    fetch(`${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`)
    .then(Response => Response.json()) // trabnsform the response to JSON data structure
    .then(result => this.setSearchTopStories(result)) // set the JSON data as a result in the local component state
    .catch(error => error) // if an error occurs
  }

  onSearchSubmit(event) {
    const {searchTerm} = this.state;
    this.fetchSearchTopStories(searchTerm)
    event.preventDefault();
  }
  
  componentDidMount() {
    const { searchTerm } = this.state
    this.fetchSearchTopStories(searchTerm)
  }

  onSearchChange(event) {
    this.setState({searchTerm: event.target.value}) // event.target.value ==> target input fild value
  }
/*
  plusPoints = (index) => {
    const newList = this.state.list
    newList[index].points += 1
    this.setState({list: newList})
  }
*/
  onDismiss(id) {
    const isNotId = item => item.objectID !== id;
    const updatedHits = this.state.result.hits.filter(isNotId);
    this.setState({
      result: {...this.state.result, hits: updatedHits}
    })
  }

  render() {
    const {result, searchTerm} = this.state;
    const page = (result && result.page) || 0; // Didn't understand this line what doing
    console.log(this.state.result)

    // if(!result) {return <Loading />;}

    return (
      <div className='page' >
      <div className='interactions'>
        <Search
          value={searchTerm} // value proparte
          onChange={this.onSearchChange} // onChange proparte
          onSubmit={this.onSearchSubmit}
        >
        Search Fucker
        </Search>
      </div>
      {result ?
        <Table
          list={result.hits} // list proparte
          onDismiss={this.onDismiss} // onDismiss proparte
          // plusPoints={this.plusPoints}
        /> :
        <Loading />
        }
        <div className='interactions'>
          <Button
            onClick={() => this.fetchSearchTopStories(searchTerm, page + 1)}
          >
            More
          </Button>
        </div>
        <Test />
      </div>
    )
  }
}

export default App;