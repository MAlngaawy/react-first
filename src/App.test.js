// This File comes for testing
/*
  Some Notes about unit testing
  1: Enzyme provides a mechanism to mount and traverse React.js component trees. This will help us get access to its own properties and state as well as its children props in order to run our assertions.
  2: toBe uses Object.is to test exact equality
  3: jest can test a plain javascript function
*/
import React from 'react';
import ReactDOM from 'react-dom';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Enzyme, {shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import App, { Search, Button, Table } from './App';


Enzyme.configure({ adapter: new Adapter() });

describe('Search', () => { // describe block is a test suite
  it('renders without crashing', () => { // it is a test cases
    const div = document.createElement('div');
    ReactDOM.render(<Search>Search</Search>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test('has a valid snapshot', () => { // Start a Test Case
    const component = renderer.create(<Search>Search</Search>); // render a component with the new renderer
    const tree = component.toJSON(); // transform it into JSON
    expect(tree).toMatchSnapshot(); // match the snapshot to the previously stored snapshot:
  });

})

describe('Button', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Button onClick = {() => {}}>Give me more</Button>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(
    <Button onClick={() => {}}>Give me more</Button>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

})

describe('Table', () => {

  const props = {
    list: [
      {title: '1', author: '1', num_comments: 1, points: 2, objectID: 'y'},
      {title: '2', author: '2', num_comments: 1, points: 2, objectID: 'z'},
    ],
    sortKey: 'TITLE',
    isSortReverse: false,
  };

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Table {...props} />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(<Table {...props} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('show two items in list', () => {
    const element = shallow(
      <Table {...props} />
    );

    expect(element.find('.table-row').length).toBe(2)
  })

})

describe('App', () => {

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
    });

  test('has a valid snapshot', () => {
    const component = renderer.create(<App />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});


// This what i found in the file
/*
test('renders learn react link', () => {
  const { getByText } = render(<App />);
  const linkElement = getByText(/learn react/i);
  expect(linkElement).toBeInTheDocument();
});*/
