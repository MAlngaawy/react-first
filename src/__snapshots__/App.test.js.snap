// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App has a valid snapshot 1`] = `
<div
  className="page"
>
  <div
    className="interactions"
  >
    <form
      onSubmit={[Function]}
    >
      <input
        onChange={[Function]}
        type="text"
        value="javaScript"
      />
      <button
        type="submit"
      >
        Search Fucker
      </button>
    </form>
  </div>
  <div
    className="loading"
  >
    <h1
      className="loading"
    >
      Waite MotherFuckr
    </h1>
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-angry fa-w-16 angry"
      data-icon="angry"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 496 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zM136 240c0-9.3 4.1-17.5 10.5-23.4l-31-9.3c-8.5-2.5-13.3-11.5-10.7-19.9 2.5-8.5 11.4-13.2 19.9-10.7l80 24c8.5 2.5 13.3 11.5 10.7 19.9-2.1 6.9-8.4 11.4-15.3 11.4-.5 0-1.1-.2-1.7-.2.7 2.7 1.7 5.3 1.7 8.2 0 17.7-14.3 32-32 32S136 257.7 136 240zm168 154.2c-27.8-33.4-84.2-33.4-112.1 0-13.5 16.3-38.2-4.2-24.6-20.5 20-24 49.4-37.8 80.6-37.8s60.6 13.8 80.6 37.8c13.8 16.5-11.1 36.6-24.5 20.5zm76.6-186.9l-31 9.3c6.3 5.8 10.5 14.1 10.5 23.4 0 17.7-14.3 32-32 32s-32-14.3-32-32c0-2.9.9-5.6 1.7-8.2-.6.1-1.1.2-1.7.2-6.9 0-13.2-4.5-15.3-11.4-2.5-8.5 2.3-17.4 10.7-19.9l80-24c8.4-2.5 17.4 2.3 19.9 10.7 2.5 8.5-2.3 17.4-10.8 19.9z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </div>
  <div
    className="interactions"
  >
    <div
      className="loading"
    >
      <h1
        className="loading"
      >
        Waite MotherFuckr
      </h1>
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-angry fa-w-16 angry"
        data-icon="angry"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 496 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zM136 240c0-9.3 4.1-17.5 10.5-23.4l-31-9.3c-8.5-2.5-13.3-11.5-10.7-19.9 2.5-8.5 11.4-13.2 19.9-10.7l80 24c8.5 2.5 13.3 11.5 10.7 19.9-2.1 6.9-8.4 11.4-15.3 11.4-.5 0-1.1-.2-1.7-.2.7 2.7 1.7 5.3 1.7 8.2 0 17.7-14.3 32-32 32S136 257.7 136 240zm168 154.2c-27.8-33.4-84.2-33.4-112.1 0-13.5 16.3-38.2-4.2-24.6-20.5 20-24 49.4-37.8 80.6-37.8s60.6 13.8 80.6 37.8c13.8 16.5-11.1 36.6-24.5 20.5zm76.6-186.9l-31 9.3c6.3 5.8 10.5 14.1 10.5 23.4 0 17.7-14.3 32-32 32s-32-14.3-32-32c0-2.9.9-5.6 1.7-8.2-.6.1-1.1.2-1.7.2-6.9 0-13.2-4.5-15.3-11.4-2.5-8.5 2.3-17.4 10.7-19.9l80-24c8.4-2.5 17.4 2.3 19.9 10.7 2.5 8.5-2.3 17.4-10.8 19.9z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </div>
  </div>
  <div>
    <input
      onChange={[Function]}
      type="text"
      value="Hello World"
    />
    <p>
      <strong>
        Output: 
      </strong>
       
      Hello World
    </p>
  </div>
</div>
`;

exports[`Button has a valid snapshot 1`] = `
<button
  className=""
  onClick={[Function]}
  type="button"
>
  Give me more
</button>
`;

exports[`Search has a valid snapshot 1`] = `
<form>
  <input
    type="text"
  />
  <button
    type="submit"
  >
    Search
  </button>
</form>
`;

exports[`Table has a valid snapshot 1`] = `
<div
  className="table"
>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a
        target="_blank"
      >
        1
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
       
      1
       
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
       
      1
       
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
       
      2
       
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        onDismiss
      </button>
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a
        target="_blank"
      >
        2
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
       
      2
       
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
       
      1
       
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
       
      2
       
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        onDismiss
      </button>
    </span>
  </div>
</div>
`;
